### DISCUSSION TITLE
`consult--jump-ensure-buffer` does not take partial previews into account

### DISCUSSION THREAD

**Discussion by Stebalien** [minad/consult#914] [Ideas]
This means grep, rg, etc. all fail to show the the relevant part of the file if the file is larger than the partial preview size.

**Comment by minad** [comment]
Yes, this is indeed not supported (yet?). We are still doing a little better with partial preview than before, where Consult completely refused to preview larger files. Patches for better preview are welcome, or maybe I will also come to this some time.

**Reply by Stebalien** [reply]
As a stop-gap, how about appending a message to the bottom of the file:

> ... truncated partial preview, open to view the full file ...

Otherwise, you're going to get a lot of bug reports.

**Reply by minad** [reply]
> As a stop-gap, how about appending a message to the bottom of the file:

Good point, already added in https://github.com/minad/consult/commit/c9e6bfefb8665a77f7a82e028ed47d4be9bf1921.

**Comment by Stebalien** [comment]
In terms of fixes:

1. It's possible to tell when text becomes "visible" with https://www.gnu.org/software/emacs/manual/html_node/elisp/Lazy-Properties.html. Effectively, set some special property on the ["this buffer is truncated" message](https://github.com/minad/consult/discussions/914#discussioncomment-7985584)  and add a hook that automatically loads more (or even just loads the rest).
2. This can also be fixed by replacing [widen](https://github.com/minad/consult/blob/c9e6bfefb8665a77f7a82e028ed47d4be9bf1921/consult.el#L1505) call with something that detects that the buffer is a partial preview.

Personally, I'm a fan of the first version because it should "just work". But it's also a bit magical.

**Reply by minad** [reply]
Sounds good in theory, but I have my doubts. While we load the rest we should also remove the text  in the beginning in order to keep resource usage in check. Otherwise we replicate the behavior of endless websites, which get longer and longer and slower and slower.

An alternative (and maybe more efficient) solution would be to tell the file preview that we want a certain offset or line number, and then load that directly. One can create separate buffers, where the offset would be part of the name, e.g., `Preview:large.log:200M`. One problem is to efficiently convert line numbers to file position without loading the whole thing. Emacs doesn't offer low-level facilities for file reading.

**Reply by Stebalien** [reply]
Yeah, I was trying to make it less complex as the preview system is already complex. But my proposal here is _effectively_ no different from just loading the entire file.

Personally, I've just raised the preview size limit (I use an explicit hotkey for all previews except buffer switching). Maybe make the size limit configurable based on the command? Or based on whether or not it's an automatic preview?

Otherwise, I think your point about finding the right part of the file is going to be the sticking point (without shelling out to something like sed, I guess.).

**Comment by minad** [comment]
Steven Allen ***@***.***> writes:

> Personally, I've just raised the preview size limit (I use an explicit hotkey for all
> previews except buffer switching). Maybe make the size limit configurable based
> on the command? Or based on whether or not it's an automatic preview?

I've not tried this, but perhaps `vertico-multiform-mode' can be used to
set the relevant variables.

> Otherwise, I think your point about finding the right part of the file is going to be
> the sticking point (without shelling out to something like sed, I guess.).

Actually, I think shelling out would be the best solution. Everything
which takes load away from Emacs is good and usually not too expensive.
See for example `marginalia--library-doc', which is also called during
an active minibuffer session.
